src/Microsoft.TemplateEngine.TemplateLocalizer.Core/TemplateStringExtractor.cs

```diff
@@ -14,125 +17,179 @@ namespace Microsoft.TemplateEngine.TemplateLocalizer.Core
+                    " because it does not match any of the rules for localizable elements: {0}",
+                    args.IdentifierPrefix + "." + elementName);
                 return;
             }
 
             JsonValueKind valueKind = element.ValueKind;
             if (valueKind == JsonValueKind.String)
             {
-                string identifier = identifierPrefix + "." + elementName;
-                if (_extractedStringIds.Contains(identifier))
-                {
-                    // This string was already included by an earlier rule, possibly with a different key. Skip.
-                    return;
-                }
-
-                // First complying rule always has the precedence.
-                TraversalRule leadingRule = complyingRules[0];
-                string finalKey = keyPrefix == null ? key : (keyPrefix + "." + key);
-                _extractedStringIds.Add(identifier);
-                _extractedStrings.Add(new TemplateString(identifier, finalKey, element.GetString() ?? string.Empty));
+                ProcessStringElement(element, elementName, localizationKey, args);
                 return;
             }
 
-            identifierPrefix = identifierPrefix + "." + elementName;
-
+            string newIdentifierPrefix = args.IdentifierPrefix + "." + elementName;
             if (valueKind == JsonValueKind.Array)
             {
-                foreach (TraversalRule rule in complyingRules)
-                {
-                    int childIndex = 0;
-                    foreach (var child in element.EnumerateArray())
-                    {
-                        string childElementName = childIndex.ToString();
-                        string childKey = (rule.KeyCreator ?? _defaultArrayKeyExtractor).CreateKey(child, childElementName, elementName, childIndex);
-
-                        TraverseJsonElements(child, childElementName, identifierPrefix, childKey, keyPrefix, rule.ChildRules);
-                        childIndex++;
-                    }
-                }
+                TraversalArgs newData = args;
+                newData.IdentifierPrefix = newIdentifierPrefix;
+                newData.Rules = complyingRules;
+                ProcessArrayElement(element, elementName, newData);
                 return;
             }
 
             if (valueKind == JsonValueKind.Object)
             {
-                keyPrefix = keyPrefix + "." + key;
+                TraversalArgs newData = args;
+                newData.IdentifierPrefix = newIdentifierPrefix;
+                newData.Rules = complyingRules;
+                newData.KeyPrefix = args.KeyPrefix + "." + localizationKey;
+                ProcessObjectElement(element, elementName, newData);
+            }
+        }
+
+        private void ProcessStringElement(JsonElement element, string elementName, string key, TraversalArgs data)
+        {
+            string identifier = data.IdentifierPrefix + "." + elementName;
```

Comment:
does it make sense to remove identifier from recursion and calculate it here from `JsonElement` (as it has parent)? 
this simplifies recursion

Reply:
We don't have access to parent element with `System.Text.Json`.

It would probably be possible if we used Newtonsoft.Json package, but I don't think it is worth the effort to switch now. Let me know if you think otherwise.

Reply:
yeah, we are moving from Newtonsoft.Json anyway.